### PR TITLE
marshal.go: updatePktSecurityParameters called with non v3 packet

### LINKED ISSUE #251: updatePktSecurityParameters called with non Version3 connection or packet
What's the cause of this error? Looking through the code, I'm not seeing much documentation as to why I'm seeing this. The SNMP device that I'm attempting to get data from is not SNMPv3, I can get specific oids from it using `Get` but doing a `WalkAll` from a root oid is causing this issue. Looking for clarification, thanks.

### PR DESCRIPTION
**Pull request by TimRots**
### marshal.go/sendOneRequest():
- Add v3 PDU Report documentation
- Add error handling for non-recoverable cases
### marshal.go/send():
-  Add validation to prevent calling ``updatePktSecurityParameters`` with non v3 packet 

Fixes  #251  #307

Signed-off-by: Tim Rots <tim.rots@protonmail.ch>
Co-Authored-by: Thomas Casteleyn <thomas.casteleyn@super-visions.com>

### CODE REVIEW

File: marshal.go [hunk 1]
```diff
@@ -295,15 +295,36 @@ func (x *GoSNMP) sendOneRequest(packetOut *SnmpPacket,
...
+			//    The request identifier of the message that triggered the report
+			//    or zero if the request identifier cannot be extracted.
+			// - The variable bindings will contain a single object identifier and its value
+			if result.Version == Version3 && result.PDUType == Report && len(result.Variables) == 1 {
 				switch result.Variables[0].Name {
-				case ".1.3.6.1.6.3.15.1.1.1.0", ".1.3.6.1.6.3.15.1.1.2.0",
-					".1.3.6.1.6.3.15.1.1.3.0", ".1.3.6.1.6.3.15.1.1.4.0",
-					".1.3.6.1.6.3.15.1.1.5.0", ".1.3.6.1.6.3.15.1.1.6.0":
+				case ".1.3.6.1.6.3.15.1.1.1.0":
+					return nil, fmt.Errorf("error: unknown security level (sec_level)")
```
Comment (Hipska) [inline marshal.go:315]: `sec_level` is a telegraf configuration name, in gosnmp `MsgFlags` is used.
```suggestion
					return nil, fmt.Errorf("error: unknown security level")
```
Same counts for the other cases below.

File: marshal.go [hunk 2]
```diff
@@ -295,15 +295,36 @@ func (x *GoSNMP) sendOneRequest(packetOut *SnmpPacket,
...
+					return nil, fmt.Errorf("error: unknown username (sec_name)")
+				case ".1.3.6.1.6.3.15.1.1.4.0":
+					// usmStatsUnknownEngineIDs wil be retransmitted
+					break waitingResponse
+				case ".1.3.6.1.6.3.15.1.1.5.0":
+					return nil, fmt.Errorf("error: wrong digest (auth_protocol, auth_password)")
+				case ".1.3.6.1.6.3.15.1.1.6.0":
+					return nil, fmt.Errorf("error: decryption error (priv_protocol, priv_password)")
 				}
 			}
```
Comment (Hipska) [inline marshal.go:329]: What about other types of reports that are received? shouldn't those be an error as well?

File: marshal.go [hunk 3]
```diff
@@ -395,8 +416,9 @@ func (x *GoSNMP) send(packetOut *SnmpPacket, wait bool) (result *SnmpPacket, err
 		}
 	}
 
-	// detect unknown engine id error and retransmit with updated engine id
-	if len(result.Variables) == 1 && result.Variables[0].Name == ".1.3.6.1.6.3.15.1.1.4.0" {
+	// v3: detect unknown engine id error and retransmit with updated engine id
+	if result.Version == Version3 && result.PDUType == Report && len(result.Variables) == 1 &&
+		result.Variables[0].Name == ".1.3.6.1.6.3.15.1.1.4.0" {
```
Comment (Hipska) [inline marshal.go:421]: What about the case of out-of-time-window above? Now you do 2 similar checks a different way.

I would have put this part inside the `if result.Version == Version3` block from line 398 above.

**Review by Hipska — CHANGES_REQUESTED** [review]

File: marshal.go (hunk 1 above)
Comment (TimRots) [reply marshal.go]: Thanks, updated as you mentioned

File: marshal.go (hunk 3 above)
Comment (TimRots) [reply marshal.go]: yup, that was not optimal. Refactored things a bit to make the check consistent with ``sendOneRequest()``

File: marshal.go (hunk 2 above)
Comment (TimRots) [reply marshal.go:343]: Added a default case to the switch to handle these as well now

**Comment by TimRots** [comment]
Thanks for the first round of review @Hipska,
can you check if the updated PR covers all your remarks?

**Review by Hipska — APPROVED** [review]
Would it be possible to return the result anyway? Other tools then can inspect it if wanted/needed.

**Comment by TimRots** [comment]
thanks for some more review @Hipska 

> Would it be possible to return the result anyway?
> Other tools then can inspect it if wanted/needed.

that can be useful indeed, will update the PR.

**Review by Hipska — APPROVED** [review]

**Comment by Hipska** [comment]
Can you link #251 to be fixed by this PR as well?

File: marshal.go [hunk 4]
```diff
@@ -355,6 +390,7 @@ func (x *GoSNMP) send(packetOut *SnmpPacket, wait bool) (result *SnmpPacket, err
 		return nil, fmt.Errorf("&GoSNMP.Conn is missing. Provide a connection or use Connect()")
 	}
 
+	// detected unknown engine id error,
```
Comment (SuperQ) [inline marshal.go:393]: This comment doesn't seem related, copy n paste error?

File: marshal.go [hunk 5]
```diff
@@ -295,15 +306,39 @@ func (x *GoSNMP) sendOneRequest(packetOut *SnmpPacket,
...
 				switch result.Variables[0].Name {
-				case ".1.3.6.1.6.3.15.1.1.1.0", ".1.3.6.1.6.3.15.1.1.2.0",
-					".1.3.6.1.6.3.15.1.1.3.0", ".1.3.6.1.6.3.15.1.1.4.0",
-					".1.3.6.1.6.3.15.1.1.5.0", ".1.3.6.1.6.3.15.1.1.6.0":
+				case usmStatsUnsupportedSecLevels:
+					return result, fmt.Errorf("error: unknown security level")
+				case usmStatsNotInTimeWindows:
 					break waitingResponse
+				case usmStatsUnknownUserNames:
+					return result, fmt.Errorf("error: unknown username")
```
Comment (SuperQ) [inline marshal.go:333]: I wonder if we should make these error messages package level exported vars, so they can be used with `errors.Is()`.

Something like
```go
var (
	ErrUnknownUsername = errors.New("unknown username")
...
```

Also while we're here, let's drop the `error: ` text from these messages.

File: marshal.go [hunk 6]
```diff
@@ -378,38 +414,39 @@ func (x *GoSNMP) send(packetOut *SnmpPacket, wait bool) (result *SnmpPacket, err
...
-			err = x.updatePktSecurityParameters(packetOut)
-			if err != nil {
-				x.logPrintf("ERROR  updatePktSecurityParameters error: %s", err)
-				return nil, err
-			}
+		// v3: check for recoverable errors
+		if result.PDUType == Report && len(result.Variables) == 1 {
+			switch result.Variables[0].Name {
+			case usmStatsNotInTimeWindows:
+				// detected out-of-time-window error,
```
Comment (SuperQ) [inline marshal.go:421]: Maybe no comment needed here?

File: marshal.go [hunk 7]
```diff
@@ -378,38 +414,39 @@ func (x *GoSNMP) send(packetOut *SnmpPacket, wait bool) (result *SnmpPacket, err
...
+					x.logPrintf("ERROR  out-of-time-window retransmit error: %s", err)
+					return result, err
+				}
 
-			result, err = x.sendOneRequest(packetOut, wait)
-			if err != nil {
-				x.logPrintf("ERROR  out-of-time-window retransmit error: %s", err)
-				return nil, err
+			case usmStatsUnknownEngineIDs:
+				// detected unknown engine id error,
```
Comment (SuperQ) [inline marshal.go:435]: Maybe no comment needed here?

**Review by SuperQ — COMMENTED** [review]
Some minor nits, otherwise great.

File: marshal.go (hunk 4 above)
Comment (TimRots) [reply marshal.go]: thx, removed

File: marshal.go (hunk 5 above)
Comment (TimRots) [reply marshal.go]: This is a nice idea :) Added as you suggested

File: marshal.go (hunk 6 above)
Comment (TimRots) [reply marshal.go]: done

**Review by SuperQ — APPROVED** [review]
LGTM, nice.

**Comment by TimRots** [comment]
Thanks for the reviews gents 👍

File: marshal.go (hunk 5 above)
Comment (Hipska) [reply marshal.go]: I like this as well. Does that mean that other tools can do this?
```go
result, err := gs.Get([]string{oid})
if err == gosnmp.ErrUnknownSecurityLevel {
    // do something related
}
```

Comment (TimRots) [reply marshal.go]: > Does that mean that other tools can do this?
> ``if err == gosnmp.ErrUnknownSecurityLevel {``

That should be the case indeed, reading up on the [docs](https://golang.org/pkg/errors/) it's mentioned to use
```
if errors.Is(err, gosnmp.ErrUnknownSecurityLevel)
```
preferable to
```
if err == gosnmp.ErrUnknownSecurityLevel
```
because the former will succeed if err wraps ``gosnmp.ErrUnknownSecurityLevel``.
